File: GitToolBox/src/main/java/zielu/gittoolbox/ui/DatePresenterImpl.java
package zielu.gittoolbox.ui;

import com.intellij.util.text.SyncDateFormat;
import java.time.ZonedDateTime;
import java.util.Date;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import zielu.gittoolbox.config.DateType;
import zielu.gittoolbox.config.GitToolBoxConfig2;
import zielu.gittoolbox.util.DateFormattingUtil;
import zielu.intellij.util.ZDateFormatUtil;

class DatePresenterImpl implements DatePresenter {
  @Nullable
  @Override
  public String format(@NotNull DateType type, @Nullable ZonedDateTime date) {
    if (date == null) {
      return null;
    } else {
      return formatImpl(type, date);
    }
  }

  private String formatImpl(DateType type, ZonedDateTime date) {
    switch (type) {
      case AUTO: return formatPrettyDate(date);
      case ABSOLUTE: return formatAbsoluteDate(date);
      case RELATIVE: return DateFormattingUtil.formatRelativeBetweenDateTimes(Date.from(date.toInstant()), new Date());
      default: return "Unknown type " + type;
    }
  }

  private String formatPrettyDate(ZonedDateTime date) {
    return ZDateFormatUtil.formatPrettyDateTime(date, ZonedDateTime.now(), getAbsoluteDateTimeFormat());
  }

  private SyncDateFormat getAbsoluteDateTimeFormat() {
    return GitToolBoxConfig2.getInstance().absoluteDateTimeStyle.getFormat();
  }

  private String formatAbsoluteDate(ZonedDateTime date) {
    return getAbsoluteDateTimeFormat().format(Date.from(date.toInstant()));
  }
}

File: GitToolBox/src/main/kotlin/zielu/gittoolbox/ui/DatePresenterImpl.kt
package zielu.gittoolbox.ui

import com.intellij.util.text.SyncDateFormat
import zielu.gittoolbox.config.DateType
import zielu.gittoolbox.util.DateFormattingUtil
import zielu.intellij.util.ZDateFormatUtil
import java.time.ZonedDateTime
import java.util.Date

internal class DatePresenterImpl
  constructor(private val gateway: DatePresenterLocalGateway) : DatePresenter {

  constructor() : this(DatePresenterLocalGatewayImpl())

  override fun format(type: DateType, date: ZonedDateTime) = formatImpl(type, date)

  private fun formatImpl(dateType: DateType, dateTime: ZonedDateTime): String {
    return when (dateType) {
      DateType.AUTO -> formatPrettyDate(dateTime)
      DateType.ABSOLUTE -> formatAbsoluteDate(dateTime)
      DateType.RELATIVE -> formatRelative(dateTime)
      DateType.HIDDEN -> ""
    }
  }

  private fun formatPrettyDate(date: ZonedDateTime): String {
    return ZDateFormatUtil.formatPrettyDateTime(date, ZonedDateTime.now(gateway.getClock()),
      getAbsoluteDateTimeFormat())
  }

  private fun getAbsoluteDateTimeFormat(): SyncDateFormat = gateway.getAbsoluteDateTimeFormat()

  private fun formatAbsoluteDate(date: ZonedDateTime): String {
    return getAbsoluteDateTimeFormat().format(Date.from(date.toInstant()))
  }

  private fun formatRelative(date: ZonedDateTime): String {
    val now = Date.from(gateway.getClock().instant())
    return DateFormattingUtil.formatRelativeBetweenDateTimes(Date.from(date.toInstant()), now)
  }
}
